### PR TITLE
spirv-fuzz: make outliner pass use additional transformations

### LINKED ISSUE #3095: spirv-fuzz: have outliner use some auxiliary transformations to handle more complex control flow
Following dicussion on #3078, the outlining transformation has been made more restrictive with respect to the control flow that it handles.

[These tests](https://github.com/KhronosGroup/SPIRV-Tools/files/3944974/disabled_tests.txt) were removed.  The outliner should be enhanced so that it can use other transformations - such as block splitting and merging - to be capable of applying outlining in the contexts captured by the tests what were removed.

### PR DESCRIPTION
**Pull request by stefanomil**
This PR changes FuzzerPassOutlineFunctions so that it uses some
transformation that make the TransformationOutlineFunction
transformation applicable in more cases. See the discussion in
#3095 for more details.

Fixes #3095 .

### CODE REVIEW

File: source/fuzz/fuzzer_pass.cpp [hunk 1]
```diff
@@ -533,5 +534,62 @@ void FuzzerPass::MaybeAddUseToReplace(
...
+    } else {
+      // The first predecessor is out of the loop.
+      maybe_preheader = &*function->FindBlock(predecessors[0]);
+    }
+
+    // |maybe_preheader| is a preheader if it branches unconditionally to
+    // the header. We also require it not to be a loop header.
+    if (maybe_preheader->terminator()->opcode() == SpvOpBranch) {
+      return maybe_preheader;
+    }
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass.cpp:571]: I believe you did not check if it is a loop header?

File: source/fuzz/fuzzer_pass.h [hunk 2]
```diff
@@ -283,6 +283,16 @@ class FuzzerPass {
       std::vector<std::pair<protobufs::IdUseDescriptor, uint32_t>>*
           uses_to_replace);
 
+  // Returns the pointer to the preheader of the loop with header |header_id|,
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass.h:286]: I believe we don't need to say "the pointer to"

File: source/fuzz/fuzzer_pass_outline_functions.cpp [hunk 3]
```diff
@@ -49,26 +49,49 @@ void FuzzerPassOutlineFunctions::Apply() {
     }
     auto entry_block = blocks[GetFuzzerContext()->RandomIndex(blocks)];
 
-    // If the entry block starts with OpPhi, try to split it.
-    if (entry_block->begin()->opcode() == SpvOpPhi) {
-      // Find the first non-OpPhi instruction.
-      opt::Instruction* non_phi_inst = nullptr;
+    // If the entry block is a loop header, we need to get or create its
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_outline_functions.cpp:52]: Could you put all the code from this line onward into a public method. And then write tests for the method. See `test/fuzz/fuzzer_pass_donate_modules_test.cpp` for example.

File: source/fuzz/fuzzer_pass.cpp (hunk 1 above)
Comment (stefanomil) [reply source/fuzz/fuzzer_pass.cpp]: Yes, I forgot to check that. Fixed

File: source/fuzz/fuzzer_pass.h (hunk 2 above)
Comment (stefanomil) [reply source/fuzz/fuzzer_pass.h]: Changed

File: source/fuzz/fuzzer_pass_outline_functions.cpp (hunk 3 above)
Comment (stefanomil) [reply source/fuzz/fuzzer_pass_outline_functions.cpp]: Done. I have created public tests for the code that adjusts an entry block or an exit block, and wrote tests for them

File: test/fuzz/fuzzer_pass_outline_functions_test.cpp [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+         %40 = OpLabel
+               OpBranchConditional %17 %41 %38
+         %41 = OpLabel
+               OpBranch %39
+         %39 = OpLabel
+               OpBranch %37
+         %38 = OpLabel
+               OpReturn
+               OpFunctionEnd
+)";
```
Comment (paulthomson) [inline test/fuzz/fuzzer_pass_outline_functions_test.cpp:279]: Missing assert?

File: test/fuzz/fuzzer_pass_outline_functions_test.cpp [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+
+  // Block 27
+  suitable_exit_block = fuzzer_pass.MaybeGetExitBlockSuitableForOutlining(
+      context->get_instr_block(27));
+  ASSERT_TRUE(suitable_exit_block);
+  ASSERT_TRUE(suitable_exit_block->GetLabel()->result_id() == 101);
+}
+}  // namespace
+}  // namespace fuzz
+}  // namespace spvtools
```
Comment (paulthomson) [inline test/fuzz/fuzzer_pass_outline_functions_test.cpp:491]: Please have a new line at the end of the file.

File: test/fuzz/fuzzer_pass_outline_functions_test.cpp (hunk 4 above)
Comment (stefanomil) [reply test/fuzz/fuzzer_pass_outline_functions_test.cpp:281]: Yes. Fixed

File: test/fuzz/fuzzer_pass_outline_functions_test.cpp (hunk 5 above)
Comment (stefanomil) [reply test/fuzz/fuzzer_pass_outline_functions_test.cpp]: Fixed

**Pull request merged**